Add Slack Directory Loader

vowelparrot opened the pull request:
Fixes linting issue from #2835 

Adds a loader for Slack Exports which can be a very valuable source of
knowledge to use for internal QA bots and other use cases.

```py
# Export data from your Slack Workspace first.
from langchain.document_loaders import SLackDirectoryLoader

SLACK_WORKSPACE_URL = "https://awesome.slack.com"

loader = ("Slack_Exports", SLACK_WORKSPACE_URL)
docs = loader.load()
```

vowelparrot committed:
Revert "Add Slack Directory Loader (#2835)"

This reverts commit a6f767ae7acd5772d7f35a81b4aad0f5d9925370.

vowelparrot committed:
Revert "Revert "Add Slack Directory Loader (#2835)""

This reverts commit 9ded3d57b84f056e5e263031ebf2e51eb387f9f2.

vowelparrot committed:
Import order

vowelparrot committed:
Merge branch 'master' into vwp/slack-directory

dev2049 commented on langchain/document_loaders/slack_directory.py:
```diff
@@ -0,0 +1,61 @@
+
+    def _get_channel_id_map(self) -> Dict[str, str]:
+        """Get a dictionary mapping channel names to their respective IDs."""
+        channels_json_path = Path(self.file_path) / "channels.json"
```
should we just store as Path in init

dev2049 commented on langchain/document_loaders/slack_directory.py:
```diff
@@ -0,0 +1,61 @@
+        channel_paths = list(Path(self.file_path).glob("*"))
+        docs = []
+        for channel_path in channel_paths:
+            if channel_path.is_dir():
```
in name of not nesting too much, could do 
```
if not channel_path.is_dir():
  continue
...
```

dev2049 commented on langchain/document_loaders/slack_directory.py:
```diff
@@ -0,0 +1,61 @@
+                    with open(json_file, encoding="utf-8") as f:
+                        messages = json.load(f)
+                    for message in messages:
+                        text = message.get("text", "")
```
maybe L42-L60 justifies a helper fn, something like `msg_json_to_doc`?

dev2049 reviewed (commented):
worth adding simple test or two?

vowelparrot committed:
Respond to feedback

vowelparrot committed:
Add test. Update interface

vowelparrot committed:
Make not stateful

vowelparrot commented:
@dev2049 refactored, added a couple tests as well.

I also changed the interface to take in a path of a zip file. Why? The user otherwise has to manually unzip the folder, and the default nesting of the export is flat, meaning the directories will all be mixed in if the user unzips it in a pre-populated directory.

This way the user can simply download a zip file and point the loader there and we wont try to load spurious `json` files
